lib/src/model/study/study_controller.dart

```diff
@@ -0,0 +1,628 @@
+
+  Timer? _startEngineEvalTimer;
+
+  Future<void> nextChapter() async {
+    if (state.hasValue) {
+      final chapters = state.requireValue.study.chapters;
+      final currentChapterIndex = chapters.indexWhere(
+        (chapter) => chapter.id == state.requireValue.study.chapter.id,
+      );
+      goToChapter(chapters[currentChapterIndex + 1].id);
+    }
+  }
+
+  Future<void> goToChapter(StudyChapterId chapterId) async {
+    state = const AsyncValue.loading();
+    state = AsyncValue.data(
+      await _fetchChapter(
+        state.requireValue.study.id,
+        chapterId: chapterId,
+      ),
+    );
+  }
+
+  Future<StudyState> _fetchChapter(
+    StudyId id, {
+    StudyChapterId? chapterId,
+  }) async {
+    final (study, pgn) = await ref
+        .read(studyRepositoryProvider)
+        .getStudy(id: id, chapterId: chapterId);
+
+    final game = PgnGame.parsePgn(pgn);
+
+    final rootComments = IList(game.comments.map((c) => PgnComment.fromPgn(c)));
+
+    final variant = study.chapter.setup.variant;
+    final orientation = study.chapter.setup.orientation;
+
+    try {
+      _root = Root.fromPgnGame(game);
+    } on PositionSetupException {
+      return StudyState(
+        variant: variant,
+        study: study,
+        currentPath: UciPath.empty,
+        isOnMainline: true,
+        root: null,
+        currentNode: StudyCurrentNode.illegalPosition(),
+        pgnRootComments: rootComments,
+        pov: orientation,
+        isLocalEvaluationAllowed: false,
+        isLocalEvaluationEnabled: false,
+        pgn: pgn,
+      );
+    }
+
+    // don't use ref.watch here: we don't want to invalidate state when the
+    // analysis preferences change
+    final prefs = ref.read(analysisPreferencesProvider);
+
+    const currentPath = UciPath.empty;
+    Move? lastMove;
+
+    final studyState = StudyState(
+      variant: variant,
+      study: study,
+      currentPath: currentPath,
+      isOnMainline: true,
+      root: _root.view,
+      currentNode: StudyCurrentNode.fromNode(_root),
+      pgnRootComments: rootComments,
+      lastMove: lastMove,
+      pov: orientation,
+      isLocalEvaluationAllowed:
+          study.chapter.features.computer && !study.chapter.gamebook,
+      isLocalEvaluationEnabled: prefs.enableLocalEvaluation,
+      pgn: pgn,
+    );
+
+    final evaluationService = ref.watch(evaluationServiceProvider);
+    if (studyState.isEngineAvailable) {
+      await evaluationService.disposeEngine();
+
+      evaluationService
+          .initEngine(
+        _evaluationContext(studyState.variant),
+        options: EvaluationOptions(
+          multiPv: prefs.numEvalLines,
+          cores: prefs.numEngineCores,
+        ),
+      )
+          .then((_) {
+        _startEngineEvalTimer = Timer(const Duration(milliseconds: 250), () {
+          _startEngineEval();
+        });
+      });
+    }
+
+    return studyState;
+  }
+
+  @override
+  Future<StudyState> build(StudyId id) async {
```

Comment:
`build` should be the first method in a notifier class, because it is its signature.